File: iosApp/iosApp/blockchain/events/commutoswap/PaymentSentEvent.swift
//
//  PaymentSentEvent.swift
//  iosApp
//
//  Created by jimmyt on 9/9/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import web3swift

/**
 Represents a [PaymentSent](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#paymentsent) event, emitted by the CommutoSwap smart contract when the buyer in a swap reports that they have sent fiat payment.
 */
class PaymentSentEvent {
    /**
     The ID of the swap for which payment has been sent.
     */
    let id: UUID
    /**
     The ID of the blockchain on which the event was emitted.
     */
    let chainID: BigUInt
    /**
     The hash of the transaction that emitted this event, as a lowercase hex string with a "0x" prefix.
     */
    let transactionHash: String
    /**
     Creates a new `PaymentSentEvent` given a web3swift `EventParserResultProtocol` containing information from a [PaymentSent](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#paymentsent) event, or returns `nil` if the passed `EventParserResultProtocol` doesn't contain information from a PaymentSent event or does not have a valid transaction hash.
     */
    init?(_ result: EventParserResultProtocol, chainID: BigUInt) {
        guard let resultID = UUID.from(data: result.decodedResult["swapID"] as? Data) else {
            return nil
        }
        id = resultID
        self.chainID = chainID
        guard var transactionHash = result.transactionReceipt?.transactionHash.toHexString().lowercased() else {
            return nil
        }
        if !transactionHash.starts(with: "0x") {
            transactionHash = "0x" + transactionHash
        }
        self.transactionHash = transactionHash
    }
    /**
     Creates a new `PaymentSentEvent` with the given `id`, `chainID` and `transactionHash`.
     */
    init(id: UUID, chainID: BigUInt, transactionHash: String) {
        self.id = id
        self.chainID = chainID
        self.transactionHash = transactionHash
    }
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/blockchain/events/commutoswap/PaymentSentEvent.kt
package com.commuto.interfacemobile.android.blockchain.events.commutoswap

import com.commuto.interfacemobile.android.contractwrapper.CommutoSwap
import java.math.BigInteger
import java.nio.ByteBuffer
import java.util.*

/**
 * Represents a [PaymentSent](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#paymentsent) event, emitted
 * by the CommutoSwap smart contract when the buyer in a swap reports that they have sent fiat payment.
 *
 * @property swapID The ID of the swap for which payment has been sent.
 * @property chainID The ID of the blockchain on which this event was emitted.
 * @property transactionHash The hash of the transaction that emitted this event, as a lowercase hex string with a "0x"
 * prefix.
 */
class PaymentSentEvent(val swapID: UUID, val chainID: BigInteger, transactionHash: String) {

    val transactionHash: String

    init {
        this.transactionHash = if (transactionHash.startsWith("0x")) {
            transactionHash.lowercase()
        } else {
            "0x${transactionHash.lowercase()}"
        }
    }

    companion object {
        /**
         * Creates a [PaymentSentEvent] from a [CommutoSwap.PaymentSentEventResponse] and a specified [chainID].
         *
         * @param event the [CommutoSwap.PaymentSentEventResponse] containing the ID of a swap for which sending payment
         * has been reported.
         * @param chainID The ID of the blockchain on which this event was emitted.
         *
         * @return A new [PaymentSentEvent] with a swap ID as a [UUID] derived from [event]'s swap ID [ByteArray] and
         * the specified [chainID], and the transaction hash specified by [event].
         */
        fun fromEventResponse(event: CommutoSwap.PaymentSentEventResponse, chainID: BigInteger): PaymentSentEvent {
            val swapIDByteBuffer = ByteBuffer.wrap(event.swapID)
            val mostSigBits = swapIDByteBuffer.long
            val leastSigBits = swapIDByteBuffer.long
            return PaymentSentEvent(UUID(mostSigBits, leastSigBits), chainID, event.log.transactionHash)
        }
    }

}